ERCS/erc-7841.md

```diff
@@ -0,0 +1,545 @@
+
+### Message Format
+
+All messages MUST follow this format:
+
+```solidity
+/// @title Metadata type
+/// @notice Metadata for a cross-chain message
+struct Metadata {
+    /// @notice The chain identifier of the source chain
+    uint32 srcChainId;
+    /// @notice The chain identifier of the destination chain
+    uint32 destChainId;
+    /// @notice The address of the sending party
+    /// @dev 32 bytes are used to encode the address. In the case of an
+    ///     Ethereum address, the last 12 bytes can be padded with zeros
+    bytes32 srcAddress;
+    /// @notice The address of the recipient
+    /// @dev 32 bytes are used to encode the address. In the case of an 
+    ///     Ethereum address, the last 12 bytes can be padded with zeros
+    bytes32 destAddress;
+    /// @notice The identifier for a cross-chain interaction session
+    /// @dev SHOULD be unique for every new cross-chain calls
+    uint128 sessionId;
+    /// @notice The message counter within an interaction session 
+    /// @dev SHOULD be unique within a session
+    /// @dev OPTIONAL for most asynchronous bridges where every message has a 
+    ///     distinct sessionId, simply set to 0 if unused
+    /// @dev E.g. In a cross-chain call: ChainA.func1 -m1-> ChainB.func2 -m2-> 
+    ///     ChainC.func3 -m3-> ChainB.func4, the subscript i in m_i is the nonce
+    uint128 nonce;
+}
+
+/// @title Message type
+/// @notice A cross-chain message
+struct Message {
+    /// @notice The message metadata 
+    Metadata metadata;
+    /// @notice Message payload encoded using RLP serialization
+    /// @dev It may be ABI-encoded function calls, info about bridged assets, 
+    ///     or arbitrary message data
+    bytes payload;
+}
+```
+
+It is RECOMMENDED that there exists a global rollup registry service that supports registration, deregistration, and efficient lookup of a rollup's chain ID. This work is outside the scope of this ERC, however. To display the sender and receiver of a `Message`, wallets and frontends MAY use [ERC-3770](./eip-3770.md) for chain-specific addresses.
+
+### Mailbox APIs
+
+Each chain SHOULD have **two canonical Mailbox contracts, one for synchronous and the other for asynchronous messaging**, responsible for managing both incoming and outgoing messages. The following APIs are RECOMMENDED to provide the minimal required functionality. Implementations of these APIs MAY include additional functions to support customized or more complex workflows.
+
+```solidity
+/// @title Mailbox contract.
+/// @notice Mailbox for sending (resp. receiving) messages to (resp. from) 
+///     other chains, standardized for messaging protocols that support 
+///     synchronous or asynchronous, or both types of message passing.
+abstract contract Mailbox {
```

Comment:
`abstract contract`s include too much implementation detail, and as such we don't generally allow them. I don't really see anything in here that couldn't be included in an interface, unless the storage requirements of `inbox` are meaningful?